---
title: "01a_AlphaSim_POP_Brandie"
author: "Julin Maloof"
date: "`r Sys.Date()`"
output: html_document
---
This script is from Brandie.  I modified it to appropriately calculate weekly survival

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Population Simulation Attempt

Rishav demoed AlphaSimR:
<https://cran.r-project.org/web/packages/AlphaSimR/index.html>

-   Rishav's demo is on his GitHub:
<https://github.com/rishavray/IPM_practice/blob/main/IPM_genetic_simulation.Rmd>

## Libraries

```{r}
library(AlphaSimR)
library(tidyverse)
logit <- brms::logit_scaled
inv_logit <- brms::inv_logit_scaled
```

## Create Founder Population

```{r}
founderPop = runMacs(
  nInd = 2000, #number of individuals to simulate 
  nChr = 10, #number of chromosomes to simulate - may not be important to match to Streps' 14
  segSites = NULL, #number of segregating sites to keep per chromosome.
  species = "GENERIC" #The GENERIC history is meant to be a reasonable all-purpose choice. It runs quickly and models a population with an effective populations size that has gone through several historic bottlenecks.
)
```

## Get germ prob estimate from Maya's data
```{r}
germprob <- read_csv("../input/WL2_2025_germ_prob.csv")
summary(germprob) #overall mean prob = 0.39

germprob %>% 
  group_by(Cross_Type) %>% 
  summarise(n=n(), meangerm=mean(prob)) #parents mean = 0.18 (lowest); F2s = highest 0.47

germprob %>% 
  filter(Cross_Type=="Parent") #WL2 prob = 0.27
```

## Define Genetic Architecture for Traits
```{r}
SP <- SimParam$new(founderPop)
#Starts the process of building a new simulation by creating a new SimParam object and assigning a founder population to the class.

traitMeans <- c(germination.logit = logit(0.18), #logit to scale it (from gene --> trait); used avg for all parents from Maya's data 
                establishment.logit = logit(0.25),
                y1surv.logit = logit(0.35),
                #alpha = 40, #Hmax = alpha -Sets the asymptote (max growth) for the Weibull model 
                beta = 7, #Hmin = beta -Changes the y intercept (the lower limit) for the Weibull model- recruit size?
                k100 = 1, # Divide by 100 to get K.  Often described as the growth rate --\> related to the slope of the line for the Weibull model 
                #delta = 0.5, #Sets the inflection point - when you shift b/t concave to convex for the Weibull model 
                flowering.logit = logit(0.2),
                fruitPerPlant = 5)

SP$addTraitA(nQtlPerChr = 10, 
             mean = traitMeans, 
             var = c(germination.logit=1,
                                 establishment.logit=1,
                                 y1surv.logit=1,
                                 beta=2,
                                 k100 = 1,
                                 flowering.logit=1,
                                 fruitPerPlant=2)^2, #genetic Var.  Numbers are in SD
             
             gamma = TRUE, #a lot of small effect loci and a few large effect loci 
             shape = 1, #shape parameter for gamma distribution 
             name=names(traitMeans))
#addTraitAG - Randomly assigns eligible QTLs for one or more additive GxE traits. If simulating more than one trait, all traits will be pleiotropic with correlated effects.
#addTraitA - Randomly assigns eligible QTLs for one or more additive traits. If simulating more than one trait, all traits will be pleiotropic with correlated additive effects. (we will keep it simple for now)
##nQtlPerChr = number of QTLs per chromosome. Can be a single value or nChr values.
##mean = a vector of desired mean genetic values for one or more traits
##var = a vector of desired genetic variances for one or more traits
##varGxE = a vector of total genotype-by-environment variances for the traits
##name = optional name for trait(s)

SP$setVarE(h2=rep(0.7, length(traitMeans))) #set heritability 
```

## Create Initial Population

```{r}
pop = newPop(founderPop, simParam = SP)
pop1 <- setPheno(pop)
```

```{r}
pheno <- pop1@pheno %>% as_tibble()
dim(pheno)
head(pheno)
summary(pheno)
```

### Convert pheno logits to probabilities and phenotypes for early traits
```{r}
pheno <- pheno %>%
  mutate(across(ends_with(".logit"), .fns = inv_logit, 
                .names = "{.col}.prob")) %>%
  rename_with(.fn = \(n) str_replace(n, "\\.logit\\.prob", "\\.prob")) %>%
  mutate(germinated = rbinom(n(), size=1, prob=germination.prob),
         established = ifelse(germinated, rbinom(n(), size=1, prob=establishment.prob), NA),
         beta,
         k = k100/100,
         k = ifelse(k <=0, min(k[k>0], na.rm = TRUE), k), # no negative k
         k = ifelse(established, k, NA), #can only have a growth rate if you establish
         beta = ifelse(beta <=0, min(beta[beta>0], na.rm = TRUE), beta), # no negative k
         fruitPerPlant = round(fruitPerPlant), # no fractional fruit
         fruitPerPlant = ifelse(fruitPerPlant < 0, 0, fruitPerPlant) # no negative fruit

) %>% select(-k100)
```

```{r}
pheno %>%
  pivot_longer(everything()) %>%
  ggplot(aes(x=value)) +
  geom_histogram() +
  facet_wrap(~name, scales="free")
```

## Try to get weekly growth data

### Model growth with the Weibull model

```{r}
weibull <- function (t, alpha, beta, k, delta) {
  result <- alpha - (alpha - beta) * exp(-(k * t)^delta)
  return(result)
}

##Example:
growth <- tibble(week = seq(0,12, 1)) %>%
  mutate(size = weibull(t = week,
                        alpha = 40, #Hmax = alpha -  Sets the asymptote (max growth)
                        beta = 0.5, #Hmin = beta - Changes the y intercept (the lower limit)
                        k = 0.01, #k - Often described as the growth rate --\> related to the slope of the line
                        delta = 0.5)) #delta - Sets the inflection point - when you shift b/t concave to convex

growth %>%  ggplot(aes(x=week, y=size)) +
  geom_line()
```

```{r}
#testing a weekly survival function that includes size
y1surv.prob <- 0.4
weeks <- 12
estab_weeks <- 3
size.prop <- 0.2
size2.prop <- 0.05 # quadratic
size <- seq(1, 20, length.out=weeks )
size.scale <- scale(size[(estab_weeks+1):weeks  ], scale=diff(range(size))) %>% as.numeric
size.scale2 <- size.scale^2
week <- seq(1:weeks)
established <- 1

x <- y1surv.prob^(1/(weeks-estab_weeks)) + # basic weekly survival
  (size.prop * size.scale + size2.prop * size.scale2) *  # the proportion of the growth rate is affected by size
  y1surv.prob^(1/(weeks-estab_weeks)) # scaling the size adjustment to the surv.prob scale

x

prod(x)
```


### Pheno weekly
```{r}
weekly_surv <- function(d, weeks, estab_weeks, size.coef=.25, size2.coef=0) {
  # function to calculate weekly survival as a function of
  # year1 survival and size
  d %>%
    mutate(weekly.surv.prob = ifelse(week <= estab_weeks, established,
                                     min(1, # make sure it is not > 1
                                     y1surv.prob^(1/(weeks-estab_weeks)) + # basic weekly survival
                                       (size.coef * size.scale + size2.coef * size.scale^2) #*  # the proportion of the growth rate is affected by size
                                       #y1surv.prob^(1/(weeks-estab_weeks))  # scaling the size adjustment to the surv.prob scale
                                     ) # min
                                     ) #ifelse
           ) %>% # mutate 
           pull(weekly.surv.prob)
}

weeks <- 12
estab_weeks <- 3
pheno_weekly <- pheno %>% 
  mutate(Indiv_ID=row_number()) %>% 
  slice(rep(1:n(), each = weeks)) %>% #duplicate each row 12 times 
  group_by(Indiv_ID) %>% 
  mutate(week=row_number(), #12 weeks for each indiv
         week_next=lead(week, order_by = week), 
         elapsed_weeks= week_next - week, #interval length in weeks
         size = if_else(germinated==1, if_else(week==1, beta, #size in week 1 is the min size (beta)
                        weibull(t = week, #otherwise use the weibull formula to calculate size
                                alpha = 40,
                                beta = beta,
                                k = k, 
                                delta = 0.5)),
                        NA) # didn't germinate
                        ) %>%
  ungroup() %>%
  mutate(size.scale = as.numeric(scale(size, scale = diff(range(size, na.rm = TRUE))))) %>% # could also try traditional scaling, but this gives range of -1 to +1
  group_by(Indiv_ID) %>%
  
  #weekly survival
  mutate(weekly.surv.prob = weekly_surv(pick(size.scale, y1surv.prob, week, established), weeks, estab_weeks), 
         alive = ifelse(established, rbinom(n(), size = 1, prob = weekly.surv.prob), established), #stay alive this week?
         alive = cummin(alive),  # if dead in a previous week, stay dead
         y1surv = min(alive)) %>%
  
  # update other traits based on death week
  mutate(size=ifelse(alive==1, size, NA), # if plant is dead, make size NA
         flowered = ifelse(sum(alive)>=8,  #only flower if alive for at least 8 weeks
           rbinom(n(), size = 1, prob = flowering.prob), NA),
         fruitPerPlant = ifelse(flowered & sum(alive)>=10, fruitPerPlant, NA)
         ) %>%
  ungroup()
```

## Extract Genotype Scores
```{r}
# Extract genotype scores for each individual
genotype_scores <- pop1@gv %>%
  as_tibble() %>%
  rownames_to_column(var = "Indiv_ID") %>%
  mutate(Indiv_ID = as.integer(Indiv_ID),
         k=k100/100,
         k = ifelse(k <=0, min(k[k>0], na.rm = TRUE), k) # no negative k
) %>%
  select(Indiv_ID, everything())

# Join genotype scores with phenotypic data
pheno_weekly_merged <- pheno_weekly %>%
  left_join(genotype_scores, by = "Indiv_ID", suffix = c("_pheno", "_geno")) %>%
  mutate(across(matches("beta|fruit"), \(x) ifelse(x < 0, min(x[x > 0], na.rm = TRUE), x)))

pheno_weekly_merged %>% 
  select(Indiv_ID, week, size, week_next, germination.logit_geno, germination.logit_pheno)

write_csv(pheno_weekly_merged, "../output/PopSim_Phenos-Genos_v4.csv")
```